LiteCore/tests/QueryTest.cc

```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
+        if ( j < arrDim ) {
+            if ( j == 0 ) ret = "\"" + string{letters[letter].first} + "\":[";
+            else
+                ret = ",";
+            if ( atTop ) {
+                ret += "\"" + concat(letters, letter, docID) + "\"";
+                letters[letter].second++;
+            } else if ( letter + 1 < letters.size() ) {
+                if ( j == 0 ) {
+                    // We are in the curly brace, and we start the letter property and
+                    // open a curly brace for the next level.
+                    ret                      = "\"" + string{letters[letter].first} + "\":[{";
+                    letters[++letter].second = 0;
+                } else {
+                    ret                      = "},{";
+                    letters[++letter].second = 0;
+                }
+            }
+        } else {
+            ret = atTop ? "]" : "}]";
+            if ( --letter >= 0 ) { letters[letter].second++; }
+        }
+        return ret;
+    }
+
+    string multiLevelArray(std::vector<std::pair<char, unsigned>>& letters, int letter, int arrDim,
+                           const string& docID) {
+        string ret{};
+        for ( string s = multiLevelArrayIter(letters, letter, arrDim, docID); !s.empty();
+              s        = multiLevelArrayIter(letters, letter, arrDim, docID) ) {
+            ret = ret + s;
+        }
+        return ret;
+    }
+
+    // build a json of nested arrays, like:
+    // {"a":[{"b":["doc001-a0-b0","doc001-a0-b1"]},{"b":["doc001-a1-b0","doc001-a1-b1"]}]}
+    // Pre-conditions: 'a' <= a && letterDim <= 'a' +26 - a
+    string buildMultiLevelArray(char a, int letterDim, int arrDim, const string& docID) {
+        std::vector<std::pair<char, unsigned>> stack;
+        for ( short i = 0; i < letterDim; ++i ) { stack.emplace_back(nextLetter(a, i), 0); }
+        int top = 0;
+        return "{" + multiLevelArray(stack, top, arrDim, docID) + "}";
+    }
+
+    vector<string> addJSONDocs(KeyStore* store, std::function<string(const string&)>& jsonBuilder, int start, int count,
+                               const string& docPrefix = "doc") {
+        vector<string>       ret;
+        constexpr size_t     bufSize = 20;
+        ExclusiveTransaction t(store->dataFile());
+        for ( int i = 0; i < count; ++i ) {
+            char docID[bufSize];
+            snprintf(docID, bufSize, "%s%03u", docPrefix.c_str(), start + i);
+            string json = jsonBuilder(docID);
+            ret.push_back(docID);
+            AddDocWithJSON(store, docID, t, json.c_str());
+        }
+        t.commit();
+        return ret;
+    }
+
+    // Pre-conditions: 'a' <= a && letterDim <= 'a' +26 - a
+    // Example output:  "doc001-a0-b0", "doc001-a0-b1", "doc001-a1-b0", "doc001-a1-b1", "doc002-a0-b0"
+    // , "doc002-a0-b1", "doc002-a1-b0", "doc002-a1-b1", ...
+    vector<string> unnestedElements(const vector<string>& docs, char a, unsigned letterDim, unsigned arrDim) {
```

Comment:
I think it could be nice to describe in the comment what the function parameters are, and what the function is actually doing (what does "unnestedElements" mean). I don't know what `letterDim`, `arrayDim`, or `a` are supposed to be.